nixos/modules/services/networking/cloudflared.nix

```diff
@@ -334,24 +369,20 @@ in
         ];
         wantedBy = [ "multi-user.target" ];
         serviceConfig = {
-          User = cfg.user;
-          Group = cfg.group;
+          RuntimeDirectory = "cloudflared-tunnel-${name}";
+          RuntimeDirectoryMode = "0400";
+          LoadCredential = [
+            "credentials.json:${tunnel.credentialsFile}"
+          ] ++ (lib.optional (certFile != null) "cert.pem:certFile");
+
           ExecStart = "${cfg.package}/bin/cloudflared tunnel --config=${mkConfigFile} --no-autoupdate run";
           Restart = "on-failure";
+          DynamicUser = true;
         };
-      })
-    ) config.services.cloudflared.tunnels;
-
-    users.users = lib.mkIf (cfg.user == "cloudflared") {
-      cloudflared = {
-        group = cfg.group;
-        isSystemUser = true;
-      };
-    };
 
-    users.groups = lib.mkIf (cfg.group == "cloudflared") {
-      cloudflared = { };
-    };
+        environment.TUNNEL_ORIGIN_CERT = lib.mkIf (certFile != null) ''%d/cert.pem'';
```

Comment:
Could you explain what does this `%d/cert.pem` does?

Reply:
Sure, see [here](https://systemd.io/CREDENTIALS/#:~:text=In%20an%20ideal,to%20specific%20credentials.). Systemd environment variables don't allow us to put the `$CREDENTIALS_DIRECTORY` variable in directly, but if we use `%d` then systemd will replace it for us.

Reply:
Cool. Totally not confusing with any other %d format specifier :)